src/Api/Response/Response.php

```diff
@@ -90,4 +101,153 @@ public function getResultsCount()
+            ? 'did-you-mean' : $this->data[self::DATA_QUERY_INFO_MESSAGE]['queryStringType'];
+        $alternativeQuery = !empty($this->data[self::DATA_QUERY_INFO_MESSAGE]['didYouMeanQuery'])
+            ? $this->data[self::DATA_QUERY_INFO_MESSAGE]['didYouMeanQuery']
+            : $this->data[self::DATA_QUERY_INFO_MESSAGE]['currentQuery'];
+        $originalQuery = !empty($this->data[self::DATA_QUERY_INFO_MESSAGE]['didYouMeanQuery'])
+            ? $this->data[self::DATA_QUERY_INFO_MESSAGE]['currentQuery']
+            : $this->data[self::DATA_QUERY_INFO_MESSAGE]['originalQuery'];
+
+        if ($type === 'corrected') {
+            return $this->translator->trans(
+                'Findologic::Template.correctedQuery',
+                [
+                    'originalQuery' => $originalQuery,
+                    'alternativeQuery' => $alternativeQuery
+                ]
+            );
+        } else if ($type === 'improved') {
+            return $this->translator->trans(
+                'Findologic::Template.improvedQuery',
+                [
+                    'originalQuery' => $originalQuery,
+                    'alternativeQuery' => $alternativeQuery
+                ]
+            );
+        } else if ($this->data[self::DATA_QUERY_INFO_MESSAGE]['currentQuery']
+            && !empty($this->data[self::DATA_QUERY_INFO_MESSAGE]['currentQuery'])) {
+            return $this->translator->trans(
+                'Findologic::Template.queryInfoMessageQuery',
+                [
+                    'query' => $this->data[self::DATA_QUERY_INFO_MESSAGE]['currentQuery'],
+                    'hits' => $this->getResultsCount()
+                ]
+            );
+        } else if (!empty($category = $this->getSelectedCategory())) {
+            return $this->translator->trans(
+                'Findologic::Template.queryInfoMessageCat',
+                [
+                    'filterName' => $category['filterName'],
+                    'cat' => $category['cat'],
+                    'hits' => $this->getResultsCount()
+                ]
+            );
+        } else if (!empty($vendor = $this->getSelectedVendor())) {
+            return $this->translator->trans(
+                'Findologic::Template.queryInfoMessageVendor',
+                [
+                    'filterName' => $vendor['filterName'],
+                    'vendor' => $vendor['vendor'],
+                    'hits' => $this->getResultsCount()
+                ]
+            );
+        } else {
+            return $this->translator->trans(
+                'Findologic::Template.queryInfoMessageDefault',
+                [
+                    'hits' => $this->getResultsCount()
+                ]
+            );
+        }
+    }
+
+    private function getSelectedCategory(): array
+    {
+        if (empty($filters = $this->getData(Response::DATA_FILTERS))) {
+            return [];
+        }
+
+        foreach ($filters as $filter) {
```

Comment:
I think the solution is well thought, but a bit too complex. You have like a loop inside of a loop which is inside of another loop. If for some reason there would be a bug in this, it would be incredibly hard to debug.
I'd suggest picking the category and vendor not out of the response, but out of the request. This way you will only need one loop for fetching category and vendor at the same time.

E.g.

```php
foreach ($queryParams as $key => $queryParam) {
    switch ($key) {
        case 'cat':
            // We have a category; Yay!
        case 'vendor':
            // Manufacturer it is!
        default:
            continue;
    }
}
```